presto-main/src/main/java/com/facebook/presto/server/AsyncPageTransportServlet.java

```diff
@@ -0,0 +1,192 @@
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_PAGE_NEXT_TOKEN;
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_PAGE_TOKEN;
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_TASK_INSTANCE_ID;
+import static com.facebook.presto.util.TaskUtils.randomizeWaitTime;
+import static com.google.common.net.HttpHeaders.CONTENT_LENGTH;
+import static com.google.common.net.HttpHeaders.CONTENT_TYPE;
+import static io.airlift.slice.SizeOf.SIZE_OF_BYTE;
+import static io.airlift.slice.SizeOf.SIZE_OF_INT;
+import static java.lang.Long.parseLong;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static javax.servlet.http.HttpServletResponse.SC_NO_CONTENT;
+
+@WebServlet(name = "AsyncPageTransportServlet", urlPatterns = {"/v2/task/*" }, asyncSupported = true)
+public class AsyncPageTransportServlet
+        extends HttpServlet
+{
+    private static final Logger log = Logger.get(AsyncPageTransportServlet.class);
+
+    private static final Duration PAGE_RESPONSE_WAIT_TIME = new Duration(300, SECONDS);
+    private static final Duration DEFAULT_MAX_WAIT_TIME = new Duration(2, SECONDS);
+
+    private final TaskManager taskManager;
+    private final Executor responseExecutor;
+    private final ScheduledExecutorService timeoutExecutor;
+
+    @Inject
+    public AsyncPageTransportServlet(
+            TaskManager taskManager,
+            @ForAsyncRpc BoundedExecutor responseExecutor,
+            @ForAsyncRpc ScheduledExecutorService timeoutExecutor)
+    {
+        this.taskManager = requireNonNull(taskManager, "taskManager is null");
+        this.responseExecutor = requireNonNull(responseExecutor, "responseExecutor is null");
+        this.timeoutExecutor = requireNonNull(timeoutExecutor, "timeoutExecutor is null");
+    }
+
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response)
+            throws ServletException, IOException
+    {
+        Duration waitTime = randomizeWaitTime(DEFAULT_MAX_WAIT_TIME);
+        long timeout = waitTime.toMillis() + PAGE_RESPONSE_WAIT_TIME.toMillis();
+
+        String requestURI = request.getRequestURI();
+        // Example:  /v2/task/{taskId}/results/{bufferId}/{token}
+        List<String> requestURIParts = Arrays.asList(requestURI.split("/"));
+
+        TaskId taskId = TaskId.valueOf(requestURIParts.get(3));
+        OutputBufferId bufferId = OutputBufferId.fromString(requestURIParts.get(5));
+        long token = parseLong(requestURIParts.get(6));
+        DataSize maxSize = DataSize.valueOf(request.getHeader(PRESTO_MAX_SIZE));
+
+        AsyncContext asyncContext = request.startAsync(request, response);
+        asyncContext.setTimeout(timeout);
+
+        // set up async listener
+        asyncContext.addListener(new AsyncListener()
+        {
+            public void onComplete(AsyncEvent event)
+                    throws IOException
+            {
```

Comment:
Are there counters for success/failure built in, or should we add our own for monitoring purposes?

Reply:
Good point,  I originally put some counter to monitor success/failure. Later I decided to remove it (1) Client will receive error code if there is failure (2) The counter is not super useful for debugging (e.g., little signal could obtained with 5/15 minutes aggregation)

Reply:
Wouldn't it be useful to help correlate client-side issues with a specific root cause (this new endpoint)?  For example, even a simple counter of number of successes or failures would help us narrow down to that.

Reply:
We keep track of those on the client side: https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/PageBufferClient.java#L111
By the counters on the client side we can figure out that something is wrong with the endpoint. Thus I don't feel particularly strong about adding more counters.